### PR TITLE
Use StatusOr for ListBuckets.

### LINKED ISSUE #1682: Implement noex version of storage::Client::ListBuckets
The signature should be something like:

```C++
class storage::noex::Client { public:
  template <typename... Options>
  StatusOr<ListBucketsReader> ListBucketsForProject(std::string const& project_id,
                                          Options&&... options);

  template <typename... Options>
  StatusOr<ListBucketsReader> ListBuckets(Options&&... options);
};
```

We need the code and a basic integration test. The existing tests should cover more complicated cases, but use your judgement.

### PR DESCRIPTION
**Pull request by coryan**
This fixes #1682.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1854)

### CODE REVIEW

File: google/cloud/storage/examples/storage_bucket_samples.cc [hunk 1]
```diff
@@ -53,8 +53,13 @@ void ListBuckets(google::cloud::storage::Client client, int& argc,
   namespace gcs = google::cloud::storage;
   [](gcs::Client client) {
     int count = 0;
-    for (gcs::BucketMetadata const& meta : client.ListBuckets()) {
-      std::cout << meta.name() << std::endl;
+    for (auto&& meta : client.ListBuckets()) {
+      if (not meta.ok()) {
+        std::cerr << "Error reading bucket list for default project"
+                  << ", status=" << meta.status();
```
Comment (devbww) [inline google/cloud/storage/examples/storage_bucket_samples.cc]: Newline?

File: google/cloud/storage/examples/storage_bucket_samples.cc [hunk 2]
```diff
@@ -75,9 +80,13 @@ void ListBucketsForProject(google::cloud::storage::Client client, int& argc,
   namespace gcs = google::cloud::storage;
   [](gcs::Client client, std::string project_id) {
     int count = 0;
-    for (gcs::BucketMetadata const& meta :
-         client.ListBucketsForProject(project_id)) {
-      std::cout << meta.name() << std::endl;
+    for (auto&& meta : client.ListBucketsForProject(project_id)) {
+      if (not meta.ok()) {
+        std::cerr << "Error reading bucket list for project " << project_id
+                  << ", status=" << meta.status();
```
Comment (devbww) [inline google/cloud/storage/examples/storage_bucket_samples.cc]: Newline?

File: google/cloud/storage/examples/storage_bucket_samples.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_bucket_samples.cc]: Fixed.

File: google/cloud/storage/examples/storage_bucket_samples.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_bucket_samples.cc]: Fixed.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**